lib/pull_preview/license.rb

```diff
@@ -0,0 +1,162 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+# 
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+# 
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+require "openssl"
+require "date"
+require "json"
+require "base64"
+
+require_relative "license/encryptor"
+require_relative "license/boundary"
+
+module PullPreview
+  class License
+    class Error < StandardError; end
+    class ImportError < Error; end
+    class ValidationError < Error; end
+
+    class << self
+      attr_reader :encryption_key
+      @encryption_key = nil
+
+      def encryption_key=(key)
+        if key && !key.is_a?(OpenSSL::PKey::RSA)
+          raise ArgumentError, "No RSA encryption key provided."
+        end
+
+        @encryption_key = key
+        @encryptor = nil
+      end
+
+      def encryptor
+        @encryptor ||= Encryptor.new(self.encryption_key)
+      end
+
+      def load_key!(key_type)
+        return if key_type == :private && @encryption_key&.private?
+        return if key_type == :public  && @encryption_key&.public?
+        # Note: public? is always true with RSA
+
+        filename = "license_key"
+        filename += ".pub" if key_type == :public
+
+        key_data = PullPreview.root_dir.join(filename).read
+        PullPreview::License.encryption_key = OpenSSL::PKey::RSA.new key_data
+      end
+
+      def import!
+        load_key! :public
+
+        data = ENV.fetch("PULLPREVIEW_LICENSE", nil)
+        if data.nil?
+          PullPreview.logger.error "Missing PULLPREVIEW_LICENSE environment variable."
+        end
+
+        data = Boundary.remove_boundary(data)
+
+        begin
+          license_json = encryptor.decrypt(data)
+        rescue Encryptor::Error
+          PullPreview.logger.error "License data could not be decrypted."
+        end
+
+        begin
+          attributes = JSON.parse(license_json)
+        rescue JSON::ParseError
+          PullPreview.logger.error "License data is invalid JSON."
+        end
+
+        license = new(attributes)
+        
+        PullPreview.logger.info "Imported license:"
+        PullPreview.logger.info license
+
+        license
+      end
+
+      def check!(command)
+        case command
+        when "github-sync"
+          if import!()&.expired?
+            PullPreview.logger.error "License expired"
+            exit 1
```

Comment:
Don't exit, raise an error instead. In CLI mode we'll catch it in bin/pullpreview